File: java-spring/src/main/java/com/example/javaspring/service/UserService.java
package com.example.javaspring.service;

import com.example.javaspring.config.BadRequestException;
import com.example.javaspring.config.DataNotFoundException;
import com.example.javaspring.db.UserEntity;
import com.example.javaspring.db.UserModel;
import com.example.javaspring.db.UserRepository;
import com.example.javaspring.service.command.CreateUser;
import com.example.javaspring.service.command.EditUser;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
public class UserService {

    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public UserModel getMe(Long userId) {
        UserEntity user = userRepository.findById(userId).orElseThrow(()->new DataNotFoundException());
        return new UserModel(user);
    }

    public UserModel getByName(String name) {
        UserEntity user = userRepository.findByName(name).orElseThrow(DataNotFoundException::new); // 위의 람다식과 동일
        return new UserModel(user);
    }

    public Page<UserModel> getUsers(Pageable pageable) {
        return userRepository.findAll(pageable).map(UserModel::new); // .map((user) -> new UserModel(user))과 동일
    }

    public UserModel createUser(CreateUser command){
        if (userRepository.findByName(command.getName()).isPresent())
            throw new BadRequestException("이미 사용되는 이름입니다.");
        UserEntity user = userRepository.save(command.toEntity());
        return new UserModel(user);
    }

    public UserModel editUser(EditUser command, Long userId){
        UserEntity user = userRepository.findById(userId).orElseThrow(DataNotFoundException::new);
        if (command.getName() == null && command.getAge() == null) // 방어로직. 불필요한 트랜잭션 작업 방지
            throw new BadRequestException("수정 사항이 없습니다.");
        if (command.getName() != null) // 수정하기 위해 값을 입력받은 필드만 변경되어 저장됨
            user.setName(command.getName());
        if(command.getAge() != null) // 일일이 null 값 여부 체크 필요
            user.setAge(command.getAge());
        // JPA 영속성 컨텍스트에 의해 값을 수정하면 최종적으로 해당 값에 부합하도록 트랜잭션 보내짐
        return new UserModel(user);
    }
}

File: kotlin-spring/src/main/kotlin/com/example/kotlinspring/service/UserService.kt
package com.example.kotlinspring.service

import com.example.kotlinspring.config.BadRequestException
import com.example.kotlinspring.config.DataNotFoundException
import com.example.kotlinspring.db.UserModel
import com.example.kotlinspring.db.UserRepository
import com.example.kotlinspring.service.command.CreateUser
import com.example.kotlinspring.service.command.EditUser
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service

@Service
class UserService(
    private val userRepository: UserRepository,
){
    fun getMe(userId: Long): UserModel =
        userRepository.findById(userId).orElseThrow(::DataNotFoundException).let {
            UserModel.of(it)
        }

    fun getByName(name: String): UserModel =
        userRepository.findByName(name).orElseThrow(::DataNotFoundException).let {
            UserModel.of(it)
        }

    fun getUsers(pageable: Pageable): Page<UserModel> =
        userRepository.findAll(pageable).map { UserModel.of(it) }

    fun createUser(command: CreateUser){
        if (userRepository.findByName(command.name).isPresent)
            throw BadRequestException("이미 사용되는 이름입니다.")
        userRepository.save(command.toEntity())
    }

    fun editUser(command: EditUser, userId: Long){
        userRepository.findById(userId).orElseThrow(::DataNotFoundException).let {
            if (command.name == it.name && command.age == it.age)
                throw BadRequestException("수정 사항이 없습니다.") // 불필요한 트랜잭션 작업 방지

            // val이기 때문에 별도의 인스턴스 생성하여 해당 엔티티 업데이트하는 트랜잭션 직접 보내기
            userRepository.save(it.copy(
                name = command.name ?: it.name, // 수정하기 위해 값을 입력받은 필드만 변경되어 저장됨
                age = command.age ?: it.age
            ))
        }
    }
}
